lib/ruby_lsp/type_checker.rb

```diff
@@ -0,0 +1,79 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyLsp
+  # A minimalistic type checker to try to resolve types that can be inferred without requiring a type system or
+  # annotations
+  class TypeChecker
+    extend T::Sig
+
+    sig { params(index: RubyIndexer::Index).void }
+    def initialize(index)
+      @index = index
+    end
+
+    sig { params(node_context: NodeContext).returns(T.nilable(String)) }
+    def infer_receiver_type(node_context)
+      node = node_context.node
+
+      case node
+      when Prism::CallNode
+        infer_receiver_for_call_node(node, node_context)
+      when Prism::InstanceVariableReadNode, Prism::InstanceVariableAndWriteNode, Prism::InstanceVariableWriteNode,
+        Prism::InstanceVariableOperatorWriteNode, Prism::InstanceVariableOrWriteNode, Prism::InstanceVariableTargetNode
+
+        return node_context.fully_qualified_name if node_context.surrounding_method
+
+        nesting = node_context.nesting
+        "#{nesting.join("::")}::<Class:#{nesting.last}>"
+      end
+    end
+
+    private
+
+    sig { params(node: Prism::CallNode, node_context: NodeContext).returns(T.nilable(String)) }
+    def infer_receiver_for_call_node(node, node_context)
+      receiver = node.receiver
+
+      case receiver
+      when Prism::SelfNode, nil
+        return node_context.fully_qualified_name if node_context.surrounding_method
+
+        # If we're not inside a method, then we're inside the body of a class or module, which is a singleton
+        # context
+        nesting = node_context.nesting
+        "#{nesting.join("::")}::<Class:#{nesting.last}>"
+      when Prism::ConstantPathNode, Prism::ConstantReadNode
+        # When the receiver is a constant reference, we have to try to resolve it to figure out the right
+        # receiver. But since the invocation is directly on the constant, that's the singleton context of that
+        # class/module
+        receiver_name = constant_name(receiver)
+        return unless receiver_name
+
+        resolved_receiver = @index.resolve(receiver_name, node_context.nesting)
+        name = resolved_receiver&.first&.name
+        return unless name
+
+        *parts, last = name.split("::")
+        return "#{last}::<Class:#{last}>" if T.must(parts).empty?
+
+        "#{T.must(parts).join("::")}::#{last}::<Class:#{last}>"
+      end
+    end
```

Comment:
We may need to look past the surrounding method to discard cases such has this:

```rb
class Foo
  def foo
    Class.new do
      bar
    end
  end
end
```

which I think would wrongly resolve to `Foo` at the moment